### ISSUE TITLE
Jakarta EE 10: Support Jakarta Concurrency 3.0

### ISSUE THREAD

**Issue by njr-11** [OpenLiberty/open-liberty#18228]
## Description of the high level feature, including any external spec links:  

Implement the Jakarta Concurrency 3.0 specification of Jakarta EE 10.
Open Liberty will add a `concurrent-3.0` feature for this.
Potential enhancements targeted for this release of the specification are shown [here](https://github.com/eclipse-ee4j/concurrency-api/milestone/9), including the following areas of focus:

- completion stages and completable futures (alignment with MicroProfile)
- propagation of third party context types
- modernization of the Trigger mechanism and cron support
- inclusion in web profile
- resource definition annotations
- asynchronous methods
- context propagation to parallel streams

Some areas originally proposed for 3.0 which have not (thus far) made it into the spec and will need to be deferred to Jakarta EE 11 or later:

- propagation of CDI context
- propagation of transaction context
- annotations from other specs (Schedule, Lock, MaxConcurrency)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/zvc0qqt937ia5kvfcqqrfizi0e241zqk
- FTS: https://github.com/openliberty/open-liberty/issues/18259
- Beta Blog Post (if applicable): https://github.com/openliberty/open-liberty/issues/20014
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/23292

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [x] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by andymc12** [comment]
UFO Review was held 2021-11-05 9am US CT - Recording is available to logged in users at https://ibm.box.com/shared/static/5rlxu8f1wo9ngf80i49ion2xumm7zcil.mp4
Notes:
- page 6 - mention Concurrency 2.0 (just package change) - why 3.0 is the "next" version after 1.0?
- page 7? - conversation on Loom (no planned support in spec or implementation) - possible no action here
- page 8 - typo "fore" should be "for"
- page 13 - @Asynchronous - why are we treating EJB separately?  or rather, why can't the Concurrency spec's @Asynchronous work for EJBs in addition to all the rest of CDI beans
	- it's different than MP FT where @Asynchronous must be either/or (Concurrency's annotation or MP FT's annotation - but both are invalid)
	- Nate to go back to community on that question - but planned spec behavior may not change.

**Comment by njr-11** [comment]
@andymc12 pages 6, 7, and 8 have been updated and I started off the discussion about asynchronous in EJB. The resolution to the latter concerns language in the Concurrency spec itself, not the Upcoming Feature Overview document, so this should be ready for approval.

**Comment by njr-11** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

    - Yes, in Upcoming Feature Overview slide 28.
    - Yes, we have covered these scenarios.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).

    a) What problem paths were tested and demonstrated?
      - Application-defined ManagedExecutorService does not become available because it's context propagation is misconfigured with the JNDI name of sopmething that is not a ContextService.
      - Application-defined ContextService is misconfigured with conflicting ways of propagating/clearing a context type.
      - Attempt to create a serializable contextual proxy from an application-defined ContextService that is configured to propagate context types that are not serializable.
      - Attempt to invalidly apply the Asynchronous annotation at class level.
      - Attempt to supply an executor JNDI name paramter to the Asynchronous annotation where the JNDI name resolves to something other than an executor.

    b) Who did you demo to?
      - Greg Watts, Jim Stephens

    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
      - Yes

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?
      - N/A

    b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
      - N/A

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
    - WAS L2: CET
    - WAS L3: EEConcurrency/AsyncBeans

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
    - No

**Comment by njr-11** [comment]
@skasund the skills transfer presentation is now complete and ready for your review:
https://ibm.box.com/s/u7fx93udkfdbl2ygwn72g8gm5kp4d190

**Comment by njr-11** [comment]
@chirp1 https://github.com/OpenLiberty/docs/issues/5725 has been opened for the ID updates

**Comment by malincoln** [comment]
@samwatibm @steven1046 @jhanders34 can you provide focal approvals?  Thanks

**Comment by samwatibm** [comment]
1) Are there translations?  
2) If so, were they won't even be sent out until tomorrow, Friday Sep 23.  (Unless they were sent in prior releases.)

**Comment by KyleAure** [comment]
@samwatibm 
1. Are there translations?
  - This feature is harder to decern if there are updated translation files.  
  - Since it was worked on by multiple developers I can't be sure which files were updated unless I go through each PR.
  - Is there a tool we can run to see outstanding translatable files?

**Comment by samwatibm** [comment]
Tagging @pnickoll in case he can help.

Translations are sent automatically so the safe thing would be to wait for 22.0.0.11 translations are back just before GM and approvals can wait until then.

**Comment by chirp1** [comment]
Nathan provided the information to doc at https://github.com/OpenLiberty/docs/issues/5725 . Approving.

**Comment by malincoln** [comment]
> Tagging @pnickoll in case he can help.
> 
> Translations are sent automatically so the safe thing would be to wait for 22.0.0.11 translations are back just before GM and approvals can wait until then.

@samwatibm are the translations back now so that the approval be given? Thanks!

**Comment by samwatibm** [comment]
220011 translations were back last month.  We are still waiting for 220012.  Approving on the basis that we do NOT need to wait for 220012.
